src/main/java/de/dhbw/karlsruhe/grammar/generation/GrammarConcatenationVerification.java

```diff
@@ -0,0 +1,135 @@
+package de.dhbw.karlsruhe.grammar.generation;
+
+import de.dhbw.karlsruhe.models.GrammarProduction;
+
+import java.util.*;
+
+public class GrammarConcatenationVerification {
+
+    public boolean verifyProductions(List<GrammarProduction> pGrammarRulesToCheck, String[] pNonTerminals) {
+        boolean valid = false;
+        valid = this.checkStartSymbolIsMappingToOtherNonTerminals(pGrammarRulesToCheck, pNonTerminals);
+
+        if (!valid) {
+            return false;
+        } else {
+            valid = this.checkLoopInSingleTerminal(pGrammarRulesToCheck);
+        }
+
+        if (!valid) {
+            return false;
+        } else {
+            valid = this.checkEveryNonTerminalIsReached(pGrammarRulesToCheck, pNonTerminals);
+        }
+
+        return valid;
+    }
+
+    public Set<String> getNonTerminatingNonTerminals(Set<GrammarProduction> pGrammarRulesToCheck, String[] pNonTerminals) {
+        Set<String> nonTerminatingNonTerminals = new HashSet<>();
+        boolean loop;
+
+        for (String currentNonTerminal : pNonTerminals) {
+            loop = this.checkIfNonTerminalLoopsItself(currentNonTerminal, pGrammarRulesToCheck);
+
+            if (loop) {
+                nonTerminatingNonTerminals.add(currentNonTerminal);
+            }
+        }
+
+        return nonTerminatingNonTerminals;
+    }
+    private boolean checkIfNonTerminalLoopsItself(String pNonTerminal, Set<GrammarProduction> pGrammarRulesToCheck) {
+        List<String> leftSidesOfUsage = new ArrayList<>();
+        String rightSidesOfNonTerminalToCheck = "";
+
+        for (GrammarProduction gr : pGrammarRulesToCheck) {
+            if (gr.leftSide().equals(pNonTerminal)) {
+                rightSidesOfNonTerminalToCheck += gr.rightSide();
+            }
+
+            if (gr.rightSide().contains(pNonTerminal)) {
+                leftSidesOfUsage.add(gr.leftSide());
+            }
+        }
+
+        if (leftSidesOfUsage.contains(pNonTerminal) || this.checkIfRightSideContainsUsedNonTerminals(rightSidesOfNonTerminalToCheck, leftSidesOfUsage)) {
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    private boolean checkIfRightSideContainsUsedNonTerminals(String rightSidesOfNonTerminalToCheck, List<String> leftSidesOfUsage) {
+        for (int i = 0; i < leftSidesOfUsage.size(); i++) {
+            if (rightSidesOfNonTerminalToCheck.contains(leftSidesOfUsage.get(i))) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private boolean checkStartSymbolIsMappingToOtherNonTerminals(List<GrammarProduction> pGrammarRulesToCheck, String[] pNonTerminals) {
+        boolean valid = false;
+
+        String prevProductionNonTerminal = "";
+        String currProductionNonTerminal = "";
+
+        String startSymbol = pGrammarRulesToCheck.get(0).leftSide();
```

Comment:
In der Generierung bestimmst du das Startsymbol über this.NonTerminals[0]. Momentan ist das noch das gleiche aber falls sich die Implementierung ändert könnte in der ersten Produktion nicht das gleiche NonTerminal stehen. ->Lösung an beiden Stellen einheitlich